core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/If.java

```diff
@@ -0,0 +1,190 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package uk.gov.gchq.gaffer.operation.impl;
+
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.core.type.TypeReference;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+import org.apache.commons.lang3.exception.CloneFailedException;
+
+import uk.gov.gchq.gaffer.operation.Operation;
+import uk.gov.gchq.gaffer.operation.Operations;
+import uk.gov.gchq.gaffer.operation.io.InputOutput;
+import uk.gov.gchq.gaffer.operation.serialisation.TypeReferenceImpl;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.function.Predicate;
+
+/**
+ * A {@code If} is an {@link Operation} which will execute one of two Operations, based on the result of testing an input Object against a provided {@link Predicate}.
+ * A simple boolean, or anything that resolves to a boolean, can also be used in place of the Predicate.
+ */
+@JsonPropertyOrder(value = {"input", "condition", "predicate", "then", "otherwise", "options"}, alphabetic = true)
+public class If implements InputOutput<Object, Object>, Operations {
+
+    private Object input;
+    private Map<String, String> options;
+    private Boolean condition;
+    private Operation then;
+    private Operation otherwise;
+
+    @JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "class")
+    private Predicate<Object> predicate;
+
+    @Override
+    public Object getInput() {
+        return input;
+    }
+
+    @Override
+    public void setInput(final Object input) {
+        this.input = input;
+    }
+
+    @Override
+    public TypeReference<Object> getOutputTypeReference() {
+        return new TypeReferenceImpl.Object();
+    }
+
+    @Override
+    public If shallowClone() throws CloneFailedException {
+        return new If.Builder()
+                .input(input)
+                .condition(condition)
+                .predicate(predicate)
+                .then(then)
+                .otherwise(otherwise)
+                .build();
+    }
+
+    @Override
+    public Map<String, String> getOptions() {
+        return options;
+    }
+
+    @Override
+    public void setOptions(final Map<String, String> options) {
+        this.options = options;
+    }
+
+    @Override
+    public Collection getOperations() {
+        return null;
```

Comment:
This needs fixing. This is required to allow things like the AddOperationsToChain and OperationChainLimiter graph hooks to work. I'm not sure what the best solution is. Perhaps something like this, but it is a bit fragile and would need a fair bit of testing:
```java
@Override
public Collection<Operation> getOperations() {
    if (null == then) {
        if (null == otherwise) {
            return Collections.emptyList();
        }
        return Lists.newArrayList(OperationChain.wrap(otherwise));
    }
    if (null == otherwise) {
        return Lists.newArrayList(OperationChain.wrap(then));
    }
    return Lists.newArrayList(OperationChain.wrap(then), OperationChain.wrap(otherwise));
}

@Override
public void updateOperations(final Collection<Operation> operations) {
    final Iterator<Operation> itr = operations.iterator();
    if (null != then) {
        if (!itr.hasNext()) {
            throw new IllegalArgumentException("Cannot update operations there are not enough operations to set 'then'");
        }
        then = itr.next();
    }
    if (null != otherwise) {
        if (!itr.hasNext()) {
            throw new IllegalArgumentException("Cannot update operations there are not enough operations to set 'otherwise'");
        }
        otherwise = itr.next();
    }
    if (itr.hasNext()) {
        throw new IllegalArgumentException("Cannot update operations there are too many operations: " + operations.size());
    }
}
```

To calculate the score of the If operation correctly, we need to calculate the score of the then and otherwise operations and take the largest score. This logic should be done by creating an implementation of ScoreResolver.